File: harness-py/tests/tools/echo_mcp_server.py
"""MCP server fixture for tests that need a real stdio server.

Exposes an ``echo`` tool and an ``add`` tool so tests can check that a server's tools reach an
agent (and its subagents) under the configured prefix.

Usage:
    $ python echo_mcp_server.py
"""

from importlib import import_module

try:
    from mcp.server import FastMCP
except ImportError:
    FastMCP = import_module("mcp.server.mcpserver").MCPServer


def start_echo_server() -> None:
    """Initialize and start the MCP server over stdio transport."""
    mcp = FastMCP("Echo Test Server")

    @mcp.tool(description="Echoes the given text back", structured_output=False)
    def echo(text: str) -> str:
        return f"echo:{text}"

    @mcp.tool(description="Adds two integers", structured_output=False)
    def add(a: int, b: int) -> int:
        return a + b

    mcp.run(transport="stdio")


if __name__ == "__main__":
    start_echo_server()

File: harness-ts/test/tools/echo-mcp-server.mjs
/**
 * MCP server fixture for tests that need a real stdio server.
 *
 * Exposes an `echo` tool and an `add` tool so tests can check that a server's tools reach an agent
 * (and its subagents) under the configured prefix. Spawned over stdio by the test.
 */

import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js'
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js'
import { z } from 'zod'

const text = (value) => ({ content: [{ type: 'text', text: value }] })

const server = new McpServer({ name: 'Echo Test Server', version: '0.0.0' })

server.registerTool('echo', { description: 'Echoes the given text back', inputSchema: { text: z.string() } }, (a) =>
  text(`echo:${a.text}`)
)
server.registerTool('add', { description: 'Adds two integers', inputSchema: { a: z.number(), b: z.number() } }, (a) =>
  text(String(a.a + a.b))
)

await server.connect(new StdioServerTransport())
